// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package supercontractserver;

import "blockchain.proto";
import "storage.proto";
import "internet.proto";

// The Supercontract service definition.
service SupercontractServer {
  // Sends a greeting
  rpc Ping (PingRequest) returns (PingReply) {}
  rpc ExecuteCall (stream Request) returns (stream Response) {}
}

// The request message containing the user's name.
message PingRequest {
  string name = 1;
}

// The response message containing the greetings
message PingReply {
  string message = 1;
}

message ExecuteRequest {
  bytes contract_key = 1;
  bytes call_id = 2;
  bytes session_id = 9;
  string file_to_call = 3;
  string function_to_call = 4;
  bytes parameters = 5;
  uint64 poex_secret_data_prefix = 6;
  uint64 sc_prepayment = 7;
  uint64 sm_prepayment = 8;
  uint32 call_mode = 10;
}

message ExecuteReturns {
  bool success = 1;
  uint64 sc_consumed = 2;
  uint64 sm_consumed = 3;
  uint64 poex_secret_data = 4;
  uint32 return_val = 5;
}

message Request {
  oneof client_message {
    ExecuteRequest request = 1;
    PaymentAmountReturn payment_amount = 2;
    HeightReturn height = 3;
    HashReturn hash = 4;
    TimestampReturn timestamp = 5;
    PublicKeyReturn public_key = 6;
    BufferReturn buffer = 7;
    InternetReadBufferReturn internet_read_buffer = 8;
    OpenConnectionReturn open_connection_status = 9;
    CloseConnectionReturn close_connection_status = 10;
    OpenFileReturn open_file_status = 11;
    WriteFileReturn write_file_status = 12;
    FileReadBufferReturn file_read_buffer = 13;
    FileOperationReturn file_operation_status = 14;
    Empty empty = 15;
    ListDirReturn entries = 16;
    PathExistReturn path_exist = 17;
    IsFIleReturn is_file = 18;
  }
}

message Response {
  oneof server_message {
    ExecuteReturns returns = 1;
    GetServicePayment get_service_payment = 2;
    OpenConnection open_connection = 5;
    ReadConnectionStream read_connection_stream = 6;
    OpenFile open_file = 7;
    WriteFileStream write_file_stream = 8;
    Flush flush = 9;
    MoveFile move_file = 10;
    RemoveFile remove_file = 11;
    GetBlockHeight get_block_height = 12;
    GetBlockHash get_block_hash = 13;
    GetBlockTime get_block_time = 14;
    GetBlockGenerationTime get_block_generation_time = 15;
    GetTransactionHash get_transaction_hash = 16;
    GetCallerPublicKey get_caller_public_key = 17;
    AddTransaction add_transaction = 18;
    GetTransactionBlockHeight get_transaction_block_height = 19;
    GetResponseTransactionHash get_response_transaction_hash = 20;
    GetTransactionContent get_transaction_content = 21;
    CloseConnection close_connection = 22;
    ReadFileStream read_file_stream = 23;
    CloseFile close_file = 24;
    ListDir list_dir = 25;
    PathExist path_exist = 26;
    IsFile is_file = 27;
    CreatDir create_dir = 28;
  }
}